File: day11/2.cpp
#include <bits/stdc++.h>
#include <chrono>
using namespace std;

map<tuple<string, bool, bool>, long long> memo;

long long dfs(map<string, vector<string>> &adj, string curr, bool fft, bool dac) {
    if (curr == "out") {
        return (fft && dac) ? 1 : 0;
    }
    long long ans = 0;

    if (memo.find(tuple(curr, fft, dac)) != memo.end()) {
        return memo[make_tuple(curr, fft, dac)];
    }

    for (string &child : adj[curr]) {
        if (child == "fft")
            ans += dfs(adj, child, true, dac);
        else if (child == "dac")
            ans += dfs(adj, child, fft, true);
        else 
            ans += dfs(adj, child, fft, dac);
    }
    return memo[make_tuple(curr, fft, dac)] = ans;
}

int main() {
    auto start = chrono::high_resolution_clock::now();
    map<string, vector<string>> adj;
    ifstream inputFile("input.txt");
    string line;
    while (getline(inputFile, line)) {
        stringstream ss(line);
        string word;
        vector<string> v;
        while (ss >> word) {
            v.push_back(word);
        }

        for (int i = 1; i < v.size(); i++) {
            adj[v[0].substr(0, v[0].size() - 1)].push_back(v[i]);
        }
    }
    cout << dfs(adj, "svr", false, false) << endl;
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end - start;
    cout << "Time taken: " <<elapsed.count() << " seconds\n";
}

File: day11/2.py
from collections import defaultdict
from functools import cache

adj = defaultdict(list)
with open("input.txt") as file:
    for line in file:
        parts = line.split()
        for i in range(1, len(parts)):
            adj[parts[0][:-1]].append(parts[i])

@cache
def dfs(curr, fft, dac):
    if curr == "out":
        return fft and dac
    ans = 0
    for child in adj[curr]:
        if child == "fft":
            ans += dfs(child, True, dac)
        elif child == "dac":
            ans += dfs(child, fft, True)
        else:
            ans += dfs(child, fft, dac)
    return ans

print(dfs("svr", False, False))